torch/fx/experimental/symbolic_shapes.py

```diff
@@ -5482,3 +5483,133 @@ def run_node(self, n: torch.fx.Node):
         result = super().run_node(n)
         rebind_unbacked(detect_fake_mode().shape_env, n, result)
         return result
+
+
+def _parse_data_dependent_expr(e):
+    def _floor(x):
+        # sympy parsing transforms something like `a // b` into something like
+        # `sympy.floor((1/b) * a)`, which we need to transform back to `a // b`
+        if isinstance(x, sympy.Mul):
+            a, b = x.args
+            if isinstance(a, sympy.Rational):
+                return FloorDiv(a.p * b, a.q)
+        return sympy.floor(x)
+
+    # find the unresolved condition on unbacked symints in the error message
+    match = re.search(r"data-dependent expression (.+) \(unhinted", e.args[0])
+
+    # parse it to a sympy expression
+    return sympy.sympify(match.group(1)).replace(sympy.floor, _floor)
+
+
+def _blame_user_code(e, frame):
+    frame_summary = traceback.FrameSummary(
+        frame.f_code.co_filename,
+        frame.f_lineno,
+        frame.f_code.co_name,
+    )
+    msg = e.args[0]
+    msg += (
+        '\n\nUser code:\n' +
+        ''.join(traceback.StackSummary.from_list([frame_summary]).format())
+    )
+    e.args = (msg,)
+
+
+class _PythonPrinter(sympy.printing.str.StrPrinter):
+    """
+    Util printer that replaces sympy symbols with their source-level names
+    and renders sympy relational operators (e.g., Eq, Ne, Ge, Le) inline
+    (i.e., as ==, !=, >, <).
```

Comment:
It's not obvious to me that this handles all of our sympy functions like PythonMod correctly